### DISCUSSION TITLE
Can I use the Buf Schema Registry (BSR) private repository with Go?

### DISCUSSION THREAD

**Discussion by ogoshikazuki** [renovatebot/renovate#26363] [Request Help]
### How are you running Renovate?

Mend Renovate hosted app on github.com

### If you're self-hosting Renovate, select which platform you are using.

None

### What is your question?

Instructions on how to use BSR's Private package with Go can be found at
https://buf.build/docs/bsr/generated-sdks/go#private

According to the above, there are two ways to authenticate
- netrc
  - Example:
```
machine proxy.corp.example.com
login jrgopher
password hunter2
```
- direct
  - Example: `GOPROXY=https://jrgopher:hunter2@proxy.corp.example.com`

There are also two ways to configure Go for private modules.
- Using `GOPRIVATE=buf.build/gen/go` (works only netrc authentication)
- Using `GOPROXY=https://buf.build/gen/go,$GOPROXY` and `GONOSUMDB=buf.build/gen/go` (works both for netrc and direct authentication)

I couldn't figure out how to do netrc authentication in renovate, so I put the following in the renovate config to use direct authentication, but it failed.
```json5
{
  customEnvVariables: {
    GOPROXY: "https://$BSR_USER:$BSR_TOKEN@$MY_DOMAIN/gen/go,https://proxy.golang.org,direct",
    GONOSUMDB: "$MY_DOMAIN",
  },
}
```
Actual values are put in $BSR_USER,$BSR_TOKEN,$MY_DOMAIN.

Here is a question below.
- Can I use netrc to set up a setup like [npmrc](https://docs.renovatebot.com/getting-started/private-packages/#add-npmrc-string-to-renovate-config)?
- Do you know why the method I practiced failed?

### Logs (if relevant)

<details><summary>Logs</summary>

```

Copy/paste the relevant log(s) here, between the starting and ending backticks

```

</details>

**Comment by rarkins** [comment]
I would go with the GOPROXY approach, with username/password embedded in the URL. I'm not sure that embedding them like `$BSR_USER` works, but if you are configuring this in a config.js file then you could use `${process.env.BSR_USER}` for example

**Reply by ogoshikazuki** [reply]
@rarkins 
GOPROXY approach needs to change environment variables, is customEnvVariables the correct way to change environment variables?
Is the following method correct for encrypting environment variables?
```json5
{
  encrypted: {
    customEnvVariables: {
      GOPROXY: "https://$BSR_USER:$BSR_TOKEN@$MY_DOMAIN/gen/go,https://proxy.golang.org,direct",
      GONOSUMDB: "$MY_DOMAIN",
    },
  }
}
```

**Reply by ogoshikazuki** [reply]
@rarkins 
I am using Mend-hosted Github App, is there any way to use Buf Schema Registry private package in this case?

**Reply by rarkins** [reply]
Custom env variables can't be used in the hosted app

**Reply by ogoshikazuki** [reply]
@rarkins 
Understood. Then it is impossible to use BSR's private repository with a hosted app.

Do you have any plans to support netrc to be able to set it up like [npmrc](https://docs.renovatebot.com/getting-started/private-packages/#add-npmrc-string-to-renovate-config)?
Or do you have plans to support other ways to use Go's private packages like BSR?

**Reply by rarkins** [reply]
I think the best solution would be for us to allow custom env variables in repo config. We likely have an existing discussion or feature request on that already

**Reply by ogoshikazuki** [reply]
That feature request would be applicable to https://github.com/renovatebot/renovate/discussions/9147. 
Thank you.
